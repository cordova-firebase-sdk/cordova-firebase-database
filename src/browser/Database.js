


var VARS_FIELD = typeof Symbol === 'undefined' ? '__vars' + Date.now() : Symbol('vars');
var SUBSCRIPTIONS_FIELD = typeof Symbol === 'undefined' ? '__subs' + Date.now() : Symbol('subscriptions');

var utils = require('cordova/utils'),
    BaseClass = require('cordova-firebase-database.BaseClass'),
    LZString = require('cordova-firebase-database.LZString');

/****************************************************************************** *
 * @name Database
 ******************************************************************************/
function Database(id, database) {
  var self = this;
  BaseClass.apply(this);

  Object.defineProperty(self, 'id', {
    value: id
  });

  Object.defineProperty(self, 'database', {
    value: database
  });

  Object.defineProperty(window.plugin.firebase, '_isReady', {
    value: false,
    writable: true,
    enumerable: false
  });
}

utils.extend(Database, BaseClass);

Database.prototype.child = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.child()', options);

  var ref = this.get(options.parentId);
  var childRef = ref.child(options.path);
  this.set(options.targetId, childRef);
  onSuccess();
};

Database.prototype.endAt = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.endAt()', options);

  var ref = this.get(options.refId);
  var query = ref.endAt(options.value, options.key);
  this.set(options.queryId, query);

  onSuccess();
};

Database.prototype.startAt = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.startAt()', options);

  var ref = this.get(options.refId);
  var query = ref.startAt(options.value, options.key);
  this.set(options.queryId, query);

  onSuccess();
};

Database.prototype.equalTo = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.equalTo()', options);

  var ref = this.get(options.refId);
  var query = ref.equalTo(options.value, options.key);
  this.set(options.queryId, query);

  onSuccess();
};

Database.prototype.limitToFirst = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.limitToFirst()', options);

  var ref = this.get(options.refId);
  var query = ref.limitToFirst(options.limit);
  this.set(options.queryId, query);

  onSuccess();
};

Database.prototype.limitToLast = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.limitToLast()', options);

  var ref = this.get(options.refId);
  var query = ref.limitToLast(options.limit);
  this.set(options.queryId, query);

  onSuccess();
};

Database.prototype.ref = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.ref()', options);

  var ref = this.database.ref(options.key);
  this.set(options.id, ref);
  onSuccess();
};

Database.prototype.updateKey = function(onSuccess, onError, args) {
  var options = args[0],
    self = this;
  console.log('[broswer] reference.remove()', options);

  var ref = this.get(options.targetId);
  ref.update(options.data).then(onSuccess).catch(onError);
};

Database.prototype.removeKey = function(onSuccess, onError, args) {
  var options = args[0],
    self = this;
  console.log('[broswer] reference.remove()', options);

  var ref = this.get(options.targetId);
  ref.remove().then(function() {
    self.delete(options.targetId);
    onSuccess();
  }).catch(onError);
};

Database.prototype.setValue = function(onSuccess, onError, args) {
  var options = args[0];
  console.log('[broswer] reference.set()', options);
  var ref = this.get(options.targetId);
  ref.set(options.data).then(onSuccess).catch(onError);
};

Database.prototype.once = function(onSuccess, onError, args) {
  var self = this,
    options = args[0];
  console.log('[broswer] query.once()', options);

  var referenceOrQuery = this.get(options.targetId);
  referenceOrQuery.once(options.eventType)
    .then(function(snapshot) {
      onSuccess({
        key: snapshot.key,
        exists: snapshot.exists(),
        exportVal: LZString.compress(JSON.stringify(snapshot.exportVal())),
        getPriority: snapshot.getPriority(),
        numChildren: snapshot.numChildren(),
        val: LZString.compress(JSON.stringify(snapshot.val())),
        toJSON: LZString.compress(JSON.stringify(snapshot.toJSON()))
      });
    })
    .catch(onError);
};

Database.prototype.orderByChild = function(onSuccess, onError, args) {
  var self = this,
    options = args[0];
  console.log('[broswer] query.orderByChild()', options);

  var referenceOrQuery = this.get(options.targetId);
  try {
    var newQuery = referenceOrQuery.orderByChild(options.path);
    self.set(options.newId, referenceOrQuery);
    onSuccess();
  } catch (e) {
    onError(e);
  }
};

Database.prototype.orderByKey = function(onSuccess, onError, args) {
  var self = this,
    options = args[0];
  console.log('[broswer] query.orderByKey()', options);

  var referenceOrQuery = this.get(options.targetId);
  try {
    var newQuery = referenceOrQuery.orderByKey();
    self.set(options.newId, referenceOrQuery);
    onSuccess();
  } catch (e) {
    onError(e);
  }
};

Database.prototype.orderByPriority = function(onSuccess, onError, args) {
  var self = this,
    options = args[0];
  console.log('[broswer] query.orderByKey()', options);

  var referenceOrQuery = this.get(options.targetId);
  try {
    var newQuery = referenceOrQuery.orderByPriority();
    self.set(options.newId, referenceOrQuery);
    onSuccess();
  } catch (e) {
    onError(e);
  }
};

Database.prototype.orderByPriority = function(onSuccess, onError, args) {
  var self = this,
    options = args[0];
  console.log('[broswer] query.orderByKey()', options);

  var referenceOrQuery = this.get(options.targetId);
  try {
    var newQuery = referenceOrQuery.orderByValue();
    self.set(options.newId, referenceOrQuery);
    onSuccess();
  } catch (e) {
    onError(e);
  }
};

Database.prototype.on = function(onSuccess, onError, args) {
  var self = this,
    options = args[0];
  console.log('[broswer] query.on()', options);

  var referenceOrQuery = this.get(options.targetId);
  var listener = referenceOrQuery.on(options.eventType, function(snapshot, key) {

    var dbInstance = window.plugin.firebase.database._DBs[self.id];

    if (dbInstance) {
      var target = dbInstance.get(options.targetId);
      if (target) {
        var args = [options.eventType];
        args.push({
          key: snapshot.key,
          exists: snapshot.exists(),
          exportVal: LZString.compress(JSON.stringify(snapshot.exportVal())),
          getPriority: snapshot.getPriority(),
          numChildren: snapshot.numChildren(),
          val: LZString.compress(JSON.stringify(snapshot.val())),
          toJSON: LZString.compress(JSON.stringify(snapshot.toJSON()))
        });

        if (key) {
          args.push(key);
        }
        target._trigger.apply(target, args);
      }
    }

  });

  this.set(options.listenerId, listener);

};

Database.prototype.off = function(onSuccess, onError, args) {
  var self = this,
    options = args[0];
  console.log('[broswer] query.off()', options);

  var referenceOrQuery = this.get(options.targetId);
  var listener;
  if (options.listenerId) {
    listener = this.get(options.listenerId);
  }

  referenceOrQuery.off(options.eventType, listener);

  this.delete(options.listenerId);

};
module.exports = Database;
