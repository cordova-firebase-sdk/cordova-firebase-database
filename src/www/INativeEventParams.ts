export interface INativeEventParams {
  args: Array<any>;

  eventType: string;

  listenerId: string;
}
