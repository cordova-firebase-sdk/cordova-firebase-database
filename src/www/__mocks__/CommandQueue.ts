export const execCmd = jest.fn((params: any): Promise<any> => {
  return Promise.resolve(params);
});
