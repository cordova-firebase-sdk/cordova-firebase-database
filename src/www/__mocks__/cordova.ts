export const exec = jest.fn((onSuccess, onError, pluginName, methodName, args): any => {
  setTimeout(onSuccess, 1);
  return {
    args,
    methodName,
    pluginName,
  };
});
