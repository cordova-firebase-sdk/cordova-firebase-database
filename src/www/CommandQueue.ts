import { exec } from "cordova";
// import { Promise } from "es6-promise";
import { nextTick, PluginBase } from "cordova-firebase-core/index";

declare let Promise: any;

/**
 * Parameters for execCmd function
 */
export interface IExecCmdParams {

  /**
   * Plugin's context
   */
  context: PluginBase;

  /**
   * Plugin's name in native side.
   * If omit this, context.id is used.
   */
  pluginName?: string;

  /**
   * Execute method name
   */
  methodName: string;

  /**
   * parameters for native side
   */
  args?: Array<any>;

  /**
   * synchronize options
   */
  execOptions?: {
    sync?: boolean;
    remove?: boolean;
  };

  [key: string]: any;
}

const commandQueue: Array<any> = [];
let isWaitMethod: string;
let isExecuting: boolean;
let executingCnt: number = 0;
const MAX_EXECUTE_CNT: number = 10;

let stopRequested: boolean = false;
window.addEventListener("unload", (): void => {
  stopRequested = true;
}, {
  once: true,
});

export const execCmd = (params: IExecCmdParams): Promise<any> => {
  params.execOptions = params.execOptions || {};
  params.args = params.args || [];
  params.pluginName = params.pluginName || params.context.id;

  // If the instance has been removed, do not execute any methods on it
  // except remove function itself.
  if (params.context.isRemoved && params.execOptions.remove) {
    console.error("[ignore]" + params.context.id + "." + params.methodName + ", because removed.");
    return Promise.resolve();
  }

  // If the overlay is not ready in native side,
  // do not execute any methods except remove on it.
  if (!params.context.isReady && !params.execOptions.remove) {
    console.error("[ignore]" + params.context.id + "." + params.methodName + ", because it's not ready.");
    return Promise.resolve();
  }

  return new Promise((resolve: (...params?: Array<any>) => void, reject: (...params?: Array<any>) => void) => {
    commandQueue.push({
      request: params,

      onSuccess: (...results: Array<any>): void => {
        // -------------------------------
        // success callback
        // -------------------------------

        // Even if the method was successful,
        // but the _stopRequested flag is true,
        // do not execute further code.
        if (!stopRequested) {
          resolve.apply(params.context, results);
        } else {
          // Page will be destroyed.
          return;
        }

        if (params.methodName === isWaitMethod) {
          isWaitMethod = null;
        }

        executingCnt--;
        nextTick(privateExec);
      },

      onError: (...results: Array<any>): void => {
        // -------------------------------
        // error callback
        // -------------------------------
        if (!stopRequested) {
          reject.apply(params.context, results);
        } else {
          // Page will be destroyed.
          return;
        }

        if (params.methodName === isWaitMethod) {
          isWaitMethod = null;
        }

        executingCnt--;
        nextTick(privateExec);
      },

    });

    // In order to execute all methods in safe,
    // the maps plugin limits the number of execution in a moment to 10.
    //
    // Note that Cordova-Android drops has also another internal queue,
    // and the internal queue drops our statement if the app send too much.
    //
    // Also executing too much statements at the same time,
    // it would cause many errors in native side, such as out-of-memory.
    //
    // In order to prevent these troubles, the maps plugin limits the number of execution is 10.
    if (isExecuting || executingCnt >= MAX_EXECUTE_CNT || isWaitMethod || commandQueue.length === 0) {
      return;
    }

    privateExec();
  });
};


const privateExec = (): void => {

  // You probably wonder why there is this code because it's already simular code at the end of the execCmd function.
  //
  // Because the commandQueue might change after the last code of the execCmd.
  // (And yes, it was occurred.)
  // In order to block surely, block the execution again.
  if (isExecuting || executingCnt >= MAX_EXECUTE_CNT || isWaitMethod || commandQueue.length === 0) {
    return;
  }

  isExecuting = true;

  // Execute some execution requests (up to 10) from the commandQueue
  let task: any;
  while (commandQueue.length > 0 && executingCnt < MAX_EXECUTE_CNT) {
    if (!stopRequested) {
      executingCnt++;
    }

    // Pick up the head one.
    task = commandQueue.shift();

    // push out normal tasks if stopRequested becomes true
    if (stopRequested && !task.request.execOptions.remove) {
      executingCnt--;
      continue;
    }

    // Some methods have to block other execution requests.
    if (task.request.execOptions.sync) {
      isWaitMethod = task.request.methodName;
      // console.log(`[sync start] ${task.request.pluginName}.${task.request.methodName}`);
      exec(task.onSuccess, task.onError, task.request.pluginName, task.request.methodName, task.request.args);
      break;
    }

    // console.log(`[start] ${task.request.pluginName}.${task.request.methodName}`);
    exec(task.onSuccess, task.onError, task.request.pluginName, task.request.methodName, task.request.args);
  }

  isExecuting = false;
};
