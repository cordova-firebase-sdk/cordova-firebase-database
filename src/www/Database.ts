import { exec } from "cordova";
import {
  App,
  BaseArrayClass,
  IAppInitializeOptions,
  loadJsPromise,
  PluginBase,
} from "cordova-firebase-core/index";
import { execCmd, IExecCmdParams } from "./CommandQueue";
import { Reference } from "./DataClasses";


declare let Promise: any;
declare let window: any;

export class Database extends PluginBase {

  protected _queue: BaseArrayClass = new BaseArrayClass();

  private _app: App;

  private _options: IAppInitializeOptions;

  private _url: string;

  private _rootRef: Reference;

  constructor(app: App, options: IAppInitializeOptions) {
    super("database");
    if (!app) {
      throw new Error("app must be FirebaseApp instance");
    }
    if (!options) {
      throw new Error("options must be passed");
    }
    if (!options.databaseURL) {
      throw new Error("options.databaseURL must be passed");
    }
    this._app = app;
    this._options = options;
    this._url = options.databaseURL.replace(/(firebaseio.com).*$/, "$1");
    if (this._url.substr(-1) !== "/") {
      this._url += "/";
    }

    this._rootRef = new Reference({
      key: null,
      parent: null,
      pluginName: this.id,
      ref: null,
      url: this._url,
    }, null);
    this._rootRef._privateInit();

    // Bubbling native events
    this._on("nativeEvent", (data: INativeEventParams): void => {
      this._rootRef._trigger.call(this._rootRef, "nativeEvent", data);
    });

    this._queue._on("insert_at", (): void => {
      if (!this._isReady) {
        return;
      }
      if (this._queue._getLength() > 0) {
        const cmd: any = this._queue._removeAt(0, true);
        if (cmd && cmd.context && cmd.methodName) {
          execCmd(cmd).then(cmd.resolve).catch(cmd.reject);
        }
      }
      if (this._queue._getLength() > 0) {
        this._queue._trigger("insert_at");
      }
    });

    // Create one new instance in native side.
    this._one("fireAppReady", (): void => {
      const onSuccess = (result?: any): void => {
        this._isReady = true;
        this._queue._trigger("insert_at");
        this._trigger("ready");
      };
      const onError = (error: any): void => {
        throw new Error(error);
      };
      exec(onSuccess, onError,
      "CordovaFirebaseDatabase", "newInstance",
      [{
        appName: this._app.name,
        id: this.id,
        options: this._options,
      }]);
    });
  }

  public get app(): App {
    return this._app;
  }


  public get url(): string {
    return this._url;
  }


  /**
   * Database.goOffline
   */
  public goOffline(): void {
    this.exec({
      context: this,
      execOptions: {
        sync: true,
      },
      methodName: "goOffline",
    });
  }

  /**
   * Database.goOnline
   */
  public goOnline(): void {
    this.exec({
      context: this,
      execOptions: {
        sync: true,
      },
      methodName: "goOnline",
    });
  }

  /**
   * Database.ref
   * @param [path] - Optional path representing the location the returned Reference will point.
   * If not provided, the returned Reference will point to the root of the Database.
   * @returns Reference
   */
  public ref(path?: string): Reference {

    let key: string = null;
    let url: string = this.url;

    if (typeof path === "string") {

      if (path === "") {
        throw new Error("First argument must not be empty string");
      }

      if (/[\.#$\[\]]/.test(path)) {
        throw new Error("First argument must be a valid firebase URL and the path can't contain \".\", \"#\", \"$\", \"[\", or \"]\".");
      }

      path = path.replace(/\/$/, "");
      key = path.replace(/^.*\//, "") || null;
      if (url.substr(0, 1) !== "/") {
        url += "/" + path;
      }
      url = url.replace(/\/+/g, "/");
      url = url.replace(/https:\//, "https://");
    }

    // Create a reference instance.
    const reference: Reference = new Reference({
      key,
      parent: null,
      pluginName: this.id,
      ref: null,
      url,
    }, this._rootRef);


    this.exec({
      args: [{
        id: reference.id,
        path,
      }],
      context: this,
      methodName: "database_ref",
    }).then((): void => {
      reference._privateInit();
    });

    return reference;
  }


  /**
   * Database.refFromURL
   * https://firebase.google.com/docs/reference/js/firebase.database.Database#refFromURL
   */
  public refFromURL(url: string): Reference {

    if (typeof url !== "string") {
      throw new Error("First argument must be string");
    }

    if (url === "") {
      throw new Error("First argument must not be empty string");
    }

    const tmp: Array<any> = this.url.match(/:\/\/(.+?.firebaseio.com)/);
    if (url.indexOf(tmp[1]) === -1) {
      throw new Error("First argument must be with the current database");
    }
    url = url.replace(/\?.*$/, "");

    const tmpUrl: string = url.replace(/https:\/\/[a-z0-9]+\.firebaseio\.com\//i, "");
    if (/[\.#$\[\]]/.test(tmpUrl)) {
      throw new Error("First argument must be a valid firebase URL and the path can't contain \".\", \"#\", \"$\", \"[\", or \"]\".");
    }


    let path: string = url.replace(/^https:\/\/.+?.firebaseio.com\/?/, "");
    path = path.replace(/\/$/, "");
    path = path.replace(/^\//, "");

    let reference: Reference = null;
    let currentUrl: string = this.url;
    let newRef: Reference = null;
    const refs: Array<Reference> = (path.split(/\//)).map((key: string) => {

      currentUrl += "/" + key;

      newRef = new Reference({
        key,
        parent: reference,
        pluginName: this.id,
        ref: null,
        url: currentUrl,
      }, this._rootRef);

      // Bubbling native events
      this._on("nativeEvent", (...parameters: Array<any>): void => {
        parameters.unshift("nativeEvent");
        newRef._trigger.apply(newRef, parameters);
      });

      this.exec({
        args: [{
          id: newRef.id,
          path,
        }],
        context: this,
        methodName: "database_ref",
      }).then((): void => {
        newRef._privateInit();
      });

      reference = newRef;

      return newRef;
    });
    //
    // this.exec({
    //   args: [{
    //     id: reference.id,
    //     path,
    //   }],
    //   context: this,
    //   methodName: "database_refFromURL",
    // }).then((): void => {
    //   reference._privateInit();
    // });

    return reference;
  }


  private exec(params: IExecCmdParams): Promise<any> {
    return new Promise((resolve: (result: any) => void, reject: (error: any) => void) => {
      params.resolve = resolve;
      params.reject = reject;
      this._queue._push(params);
    });
  }

}



class SecretDatabaseManager {

  public _databases: any = {};

}

if (window.cordova && window.cordova.version) {
  const manager: SecretDatabaseManager = new SecretDatabaseManager();

  Object.defineProperty(window.plugin.firebase, "database", {
    value: (app?: App): Database => {
      if (!app) {
        // Obtains default app
        app = window.plugin.firebase.app();
      }

      // If database is created for the app, returns it.
      // Otherwise, create a new instance.
      let database: Database = manager._databases[app.name];
      if (!database) {
        database = new Database(app, app.options);
        manager._databases[app.name] = database;
      }
      return database;
    },
  });

  Object.defineProperty(window.plugin.firebase.database, "_nativeCallback", {
    enumerable: false,
    value: (dbId: string, listenerId: string, eventType: string, args: Array<any>): void => {

      const dbInstance = window.plugin.firebase.database._DBs[dbId];

      if (dbInstance) {
        dbInstance._trigger("nativeEvent", {
          args,
          eventType,
          listenerId,
        });
      }
    },
  });
}
