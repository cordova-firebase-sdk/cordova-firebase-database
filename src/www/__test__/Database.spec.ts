import { Database } from "../Database";
import { App } from "cordova-firebase-core/index";
import { exec } from "../__mocks__/cordova";
import { execCmd } from "../__mocks__/CommandQueue";
import { INativeEventParams } from "../INativeEventParams";

declare let cordova: any;

describe("[Database]", () => {

  cordova.require = jest.fn();
  cordova.define.moduleMap["cordova-firebase-database.Database"] = 1;

  const commonApp: App = new App("dummyApp", {
    hello: "world",
    databaseURL: "https://dummy.firebaseio.com/"
  });

  const commonDb: Database = new Database(commonApp, commonApp.options);
  commonDb._trigger("fireAppReady");

  beforeEach(() => {
    // Clear all instances and calls to constructor and all methods:
    exec.mockClear();
    execCmd.mockClear();
    cordova.require.mockClear();
  });

  describe("constructor", () => {
    it("should create a new instance in native side", (done) => {

      const app: App = new App("hello", {
        hello: "world",
        databaseURL: "https://dummy.firebaseio.com/"
      });
      app._one("ready", () => {
        const db: Database = new Database(app, app.options);
        db._trigger("fireAppReady");

        db._one("ready", () => {
          expect(db.app).toBe(app);
          expect(exec).toHaveBeenCalledTimes(2);
          expect(exec.mock.calls[0][2]).toBe("CordovaFirebaseCore");
          expect(exec.mock.calls[0][3]).toBe("newInstance");
          expect(exec.mock.calls[0][4][0].name).toBe("hello");
          expect(exec.mock.calls[0][4][0].options.databaseURL).toEqual("https://dummy.firebaseio.com/");

          expect(exec.mock.calls[1][2]).toBe("CordovaFirebaseDatabase");
          expect(exec.mock.calls[1][3]).toBe("newInstance");
          expect(exec.mock.calls[1][4][0].appName).toBe("hello");
          expect(exec.mock.calls[1][4][0].id).toBe(db.id);
          expect(exec.mock.calls[1][4][0].options.databaseURL).toEqual("https://dummy.firebaseio.com/");

          done();
        });
      });
    });

    it("should catch Error if something happends in native side", () => {
      exec.mockImplementationOnce(() => {
        throw new Error("Something happends!");
      });
      expect(() => {
        const app: App = new App("hello", {
          hello: "world",
          databaseURL: "https://dummy.firebaseio.com/"
        });
        const database: Database = app.database();
      }).toThrowErrorMatchingSnapshot();
    });


    it("should throw Error if app is null", () => {
      expect(() => {
        new Database(null, null);
      }).toThrowErrorMatchingSnapshot();
    });

    it("should throw Error if option is null", () => {
      expect(() => {
        const app: App = new App("hello", {
          hello: "world",
          databaseURL: "https://dummy.firebaseio.com/"
        });
        new Database(app, null);
      }).toThrowErrorMatchingSnapshot();
    });

    it("should throw Error if option does not contain 'databaseURL'", () => {
      expect(() => {
        const app: App = new App("hello", {
          hello: "world",
          databaseURL: "https://dummy.firebaseio.com/"
        });
        new Database(app, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it("should 'nativeEvent' on rootRef if database receives 'nativeEvent'", (done) => {
      const app: App = new App("hello", {
        hello: "world",
        databaseURL: "https://dummy.firebaseio.com/"
      });
      const database: Database = app.database();
      database.ref().root._one("nativeEvent", (params: INativeEventParams) => {
        expect(params.args[0]).toBe("Hello");
        expect(params.eventType).toBe("test");
        expect(params.listenerId).toBe("listenerId");
        done();
      });
      database._trigger("nativeEvent", {
        args: ["Hello"],
        eventType: "test",
        listenerId: "listenerId",
      });
    });
    //
    it("methods should work even if native side is not ready.", (done) => {

      const app: App = new App("hello", {
        hello: "world",
        databaseURL: "https://dummy.firebaseio.com/"
      });
      const database: Database = app.database();

      database.ref("users").set("myName").then(() => {
        expect(execCmd).toHaveBeenCalledTimes(2);
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("reference_set");
        done();
      });
      setTimeout(() => {
        app._trigger("ready");
      }, 10);
    });
    it("methods should work even if native side is not ready.", (done) => {

      const app: App = new App("hello", {
        hello: "world",
        databaseURL: "https://dummy.firebaseio.com/"
      });
      const database: Database = app.database();
      const ref1 = database.ref("users1");
      const ref2 = database.refFromURL("https://dummy.firebaseio.com/users2");

      ref1.set("value1");
      ref2.set("value2");

      app._trigger("ready");
      setTimeout(() => {
        expect(execCmd).toHaveBeenCalledTimes(4);
        expect(execCmd.mock.calls[0][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[1][0].methodName).toBe("database_ref");
        expect(execCmd.mock.calls[2][0].methodName).toBe("reference_set");
        expect(execCmd.mock.calls[3][0].methodName).toBe("reference_set");
        expect(ref1.toString()).toBe("https://dummy.firebaseio.com/users1");
        expect(ref2.toString()).toBe("https://dummy.firebaseio.com/users2");
        done();
      }, 30);
    });

    it("window.plugin.firebase.database() should return the default database.", () => {
      const app: App = (window as any).plugin.firebase.initializeApp({
        databaseURL: "https://dummy.firebaseio.com/",
      });
      app._trigger("ready");

      const database: Database = (window as any).plugin.firebase.database();
      expect(database.app.name).toBe("[DEFAULT]");
    });
    it("should receive nativeEvent through _nativeCallback.", (done) => {

      const database: Database = (window as any).plugin.firebase.database();
      expect(database.app.name).toBe("[DEFAULT]");

      database.ref()._one("listenerId", (params: INativeEventParams) => {
        expect(params.eventType).toBe("testEvent");
        expect(params.args[0]).toBe("HELLO WORLD");
        done();
      });
      (window as any).plugin.firebase.database._nativeCallback(database.id, "listenerId", "testEvent", ["HELLO WORLD"]);
    });
    it("getSelf() should return the database itself", () => {
      const database: Database = (window as any).plugin.firebase.database();
      expect(database.getSelf()).toBe(database);
    });
  });

  describe(".goOffline", () => {
    it("should involve native code with correct parameters.", (done) => {

      commonDb.goOffline();
      setTimeout(() => {
        expect(execCmd).toHaveBeenCalledTimes(1);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.methodName).toEqual("goOffline");
        done();
      }, 5);
    });
  });
  describe(".goOnline", () => {
    it("should involve native code with correct parameters.", (done) => {

      commonDb.goOnline();
      setTimeout(() => {
        expect(execCmd).toHaveBeenCalledTimes(1);
        const params: any = execCmd.mock.calls[0][0];
        expect(params.methodName).toEqual("goOnline");
        done();
      }, 5);
    });
  });
  describe(".ref", () => {
    it("should create correct references.", (done) => {

      const ref = commonDb.ref("users/user01");
      expect(ref.toString()).toEqual("https://dummy.firebaseio.com/users/user01");
      expect(ref.parent.toString()).toEqual("https://dummy.firebaseio.com/users");
      expect(ref.parent.parent.toString()).toEqual("https://dummy.firebaseio.com/");
      done();
    });

    it(".root.toString() should be the same as the current database", () => {
      expect(commonDb.ref().toString()).toBe("https://dummy.firebaseio.com/");
      expect(commonDb.ref().root.toString()).toBe("https://dummy.firebaseio.com/");
    });

    it(".root.key should be null", () => {
      expect(commonDb.ref().root.key).toBe(null);
    });

    it(".root.parent should be null", () => {
      expect(commonDb.ref().root.parent).toBe(null);
    });

    it("should reject invalid path", () => {
      expect(() => {
        commonDb.ref("hello/../world/");
      }).toThrowErrorMatchingSnapshot();
    });
  });
  describe(".refFromURL", () => {

    it("should create correct references.", () => {

      const ref = commonDb.refFromURL("https://dummy.firebaseio.com/users/test?awrewqar=www");
      expect(ref.toString()).toEqual("https://dummy.firebaseio.com/users/test");
      expect(ref.key).toEqual("test");
      expect(ref.parent.key).toEqual("users");
      expect(ref.parent.toString()).toEqual("https://dummy.firebaseio.com/users");
      expect(ref.parent.parent.toString()).toEqual("https://dummy.firebaseio.com/");
    });

    it("should reject empty string", () => {
      expect(() => {
        commonDb.refFromURL("");
      }).toThrowErrorMatchingSnapshot();
    });

    it("should reject invalid url", () => {
      expect(() => {
        commonDb.refFromURL("http://hello.com/users/../world/");
      }).toThrowErrorMatchingSnapshot();
    });
    it("should reject invalid url", () => {
      expect(() => {
        commonDb.refFromURL("https://dummy.firebaseio.com/users/#/world/");
      }).toThrowErrorMatchingSnapshot();
    });
    it("should reject if hostname does not match with the currend database", () => {
      expect(() => {
        commonDb.refFromURL("http://hello.com/users/../world/");
      }).toThrowErrorMatchingSnapshot();
    });
    it("should reject if url is null", () => {
      expect(() => {
        commonDb.refFromURL(null);
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
