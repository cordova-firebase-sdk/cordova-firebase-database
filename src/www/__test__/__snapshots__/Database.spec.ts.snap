// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Database] .ref should reject empty string 1`] = `"First argument must not be empty string"`;

exports[`[Database] .ref should reject invalid path 1`] = `"First argument must be a valid firebase URL and the path can't contain \\".\\", \\"#\\", \\"$\\", \\"[\\", or \\"]\\"."`;

exports[`[Database] .refFromURL should reject empty string 1`] = `"First argument must not be empty string"`;

exports[`[Database] .refFromURL should reject if hostname does not match with the currend database 1`] = `"First argument must be with the current database"`;

exports[`[Database] .refFromURL should reject invalid url 1`] = `"First argument must be with the current database"`;
